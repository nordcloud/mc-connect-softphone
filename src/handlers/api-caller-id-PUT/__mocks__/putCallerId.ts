export const putCallerId = jest
  .fn()
  .mockResolvedValue({ expiresTimestampSeconds: 1234567890 });
