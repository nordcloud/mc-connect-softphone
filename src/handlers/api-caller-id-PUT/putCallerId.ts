import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';

import { CALLER_ID_EXPIRES_SECONDS, CALLER_ID_TABLE } from '../../consts';
import { User } from '../../types';

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({}));

type Props = {
  user: User;
  callerId: string;
};

export async function putCallerId({ user, callerId }: Props) {
  const expiresTimestampSeconds =
    Math.floor(Date.now() / 1000) + CALLER_ID_EXPIRES_SECONDS;

  const command = new PutCommand({
    TableName: CALLER_ID_TABLE,
    Item: {
      username: user.email,
      callerId,
      expires: expiresTimestampSeconds,
    },
  });

  await ddbDocClient.send(command);

  return { expiresTimestampSeconds };
}
