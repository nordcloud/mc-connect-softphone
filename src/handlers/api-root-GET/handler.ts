import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { deleteCallerId } from '../../utils/deleteCallerId';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  await deleteCallerId(user.email);

  const customers = await getCustomers();

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers,
  });
};
