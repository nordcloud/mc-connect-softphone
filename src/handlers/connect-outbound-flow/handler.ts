import { ConnectContactFlowHandler } from 'aws-lambda';

import { deleteCallerId } from '../../utils/deleteCallerId';

export const handler: ConnectContactFlowHandler = async (event) => {
  const username = event.Details.Parameters.Username;

  if (!username) {
    throw new Error('Missing Username');
  }

  const callerId = await deleteCallerId(username);

  return {
    caller_id: callerId || null,
  };
};
