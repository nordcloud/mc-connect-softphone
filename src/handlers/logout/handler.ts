import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getCookie } from '../../utils/getCookie';
import { getStageConsts } from '../../utils/getStageConsts';

const { OAUTH_BASE_URL } = getStageConsts();

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { API_URL } = process.env;

  if (!API_URL) {
    throw new Error('Missing API_URL');
  }

  const idToken = getCookie(event, ID_TOKEN_COOKIE);
  let location = `${OAUTH_BASE_URL}/logout?id_token_hint=${idToken}&post_logout_redirect_uri=${API_URL}`;

  if (!idToken) {
    console.log('Missing idToken cookie');
    location = '/';
  }

  console.log(`Redirecting to ${location}`);

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location },
    cookies: [cookie.serialize(ID_TOKEN_COOKIE, '')],
  };
};
