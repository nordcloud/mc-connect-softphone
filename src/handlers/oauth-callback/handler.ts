import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getStage } from '../../utils/getStage';
import { fetchIdToken } from './fetchIdToken';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const responseCookies = [];
  const idTokenData = await fetchIdToken(event);

  if (idTokenData) {
    responseCookies.push(
      cookie.serialize(ID_TOKEN_COOKIE, idTokenData.id_token, {
        secure: getStage() !== 'local',
        httpOnly: true,
        expires: new Date(Date.now() + idTokenData.expires_in * 1000),
      })
    );
  }

  console.log('Redirecting to "/"');

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: '/' },
    cookies: responseCookies,
  };
};
