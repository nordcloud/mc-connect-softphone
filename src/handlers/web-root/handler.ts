import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers: await getCustomers(event),
  });
};
