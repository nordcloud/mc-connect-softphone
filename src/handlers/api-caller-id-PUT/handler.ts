import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import { StatusCodes } from 'http-status-codes';

import { getUser } from '../../utils/getUser';
import { putCallerId } from './putCallerId';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return { statusCode: StatusCodes.UNAUTHORIZED };
  }

  const callerId = event.body || '';
  const { expiresTimestampSeconds } = await putCallerId({ user, callerId });

  return {
    statusCode: StatusCodes.OK,
    body: JSON.stringify({ expiresTimestampSeconds }),
  };
};
