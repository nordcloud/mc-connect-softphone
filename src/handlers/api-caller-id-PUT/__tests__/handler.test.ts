import { getUser } from '../../../utils/getUser';
import { handler } from '../handler';
import { putCallerId } from '../putCallerId';

jest.mock('../../../utils/getUser');
jest.mock('../putCallerId');

const args = [{ body: 'dummyEventBody' }] as unknown as Parameters<typeof handler>;

describe('api-caller-id-PUT', () => {
  it('calls getUser with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(getUser).mock.calls).toMatchSnapshot();
  });

  it('calls putCallerId with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(putCallerId).mock.calls).toMatchSnapshot();
  });

  it('calls putCallerId with correct args when event body is missing', async () => {
    const argsClone = structuredClone(args);
    delete argsClone[0].body;
    await handler(...argsClone);
    expect(jest.mocked(putCallerId).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  it('resolves with a correct value when user is undefined', () => {
    jest.mocked(getUser).mockResolvedValueOnce(undefined);
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });
});
