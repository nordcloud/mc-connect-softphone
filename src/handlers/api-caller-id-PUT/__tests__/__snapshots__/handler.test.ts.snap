// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-caller-id-PUT calls getUser with correct args 1`] = `
[
  [
    {
      "body": "dummyEventBody",
    },
  ],
]
`;

exports[`api-caller-id-PUT calls putCallerId with correct args 1`] = `
[
  [
    {
      "callerId": "dummyEventBody",
      "user": {
        "email": "dummyEmail",
        "name": "dummyName",
        "sub": "dummySub",
      },
    },
  ],
]
`;

exports[`api-caller-id-PUT calls putCallerId with correct args when event body is missing 1`] = `
[
  [
    {
      "callerId": "",
      "user": {
        "email": "dummyEmail",
        "name": "dummyName",
        "sub": "dummySub",
      },
    },
  ],
]
`;

exports[`api-caller-id-PUT resolves with a correct value 1`] = `
{
  "body": "{"expiresTimestampSeconds":1234567890}",
  "statusCode": 200,
}
`;

exports[`api-caller-id-PUT resolves with a correct value when user is undefined 1`] = `
{
  "statusCode": 401,
}
`;
